File: leetcode/35.search-insert-position.py
from typing import List

#
# @lc app=leetcode id=35 lang=python3
#
# [35] Search Insert Position
#

# @lc code=start
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)-1

        while (right - left) > 1:
            middle = left + int((right - left)/2)

            if target >= nums[middle]: 
                left = middle
            else: 
                right = middle
        
        if target > nums[right]:
            return right+1
        
        if target > nums[left]:
            return left + 1
        
        return left

# @lc code=end



inputs = [
    [
       [1,3,5,6],
       5,
       2
    ],
    [
       [1,3,5,6],
       2,
       1
    ],
    [
       [1,3,5,6],
       7,
       4
    ],
    [
       [1,3,5,6],
       0,
       0
    ],
]
s = Solution()
for [arr, target, expected] in inputs:
    res = s.searchInsert(arr, target)
    print(res == expected, res)

File: leetcode/35.search-insert-position.go
/*
 * @lc app=leetcode id=35 lang=golang
 *
 * [35] Search Insert Position
 */

package main

import (
"fmt"
)

// @lc code=start
func searchInsert(nums []int, target int) int {
	l := 0
	r := len(nums)-1

	for l<=r {
		m := int((l+r)/2)

		switch{
		case nums[m] < target:
			l = m+1
		case nums[m] > target:
			r = m-1
		default: 
			return m
		}
	}

	return l
}
// @lc code=end


func main() {
	tests := []struct {
		arr     []int
		target  int
		out     int
	}{
		{[]int{1,3,5,5,5,6}, 5, 2},
		{[]int{1,3,5,6}, 5, 2},
		// {[]int{1,3,3,5,5,5,6}, 2, 1},
		// {[]int{1,3,3,5,5,5,6}, 4, 3},
		// {[]int{5,5,5}, 5, 0},
		// {[]int{1,5,5,5}, 2, 1},
		// {[]int{1,3,5,6}, 2, 1},
		// {[]int{1,3,5,6}, 7, 4},
		// {[]int{1,3,5,6}, 0, 0},
		// {[]int{1,3}, 2, 1},
	}

	for _, test := range tests {
		res := searchInsert(test.arr, test.target)
		fmt.Println(res == test.out, res)
	}
}
